restord/ImageSessionServer.m

```diff
@@ -110,34 +110,61 @@ - (void)beginImaging {
   NSLog(@"%@ ASR exit code: %d", self, asrReturnCode);
 
-  NSError *err;
-  if (asrReturnCode == 0) {
-    DADiskRef disk = NULL;
-
-    // Get the synthesized "Macintosh HD" apfs disk, if any.
-    NSString *apfsOSDisk = [self apfsOSDisk];
-    if (apfsOSDisk) {
-      NSString *apfsBootDisk = [NSString stringWithFormat:@"/dev/%@", apfsOSDisk];
-      disk = DADiskCreateFromBSDName(NULL, self.diskArbSession, apfsBootDisk.UTF8String);
-    }
-
-    NSURL *mountURL = [self mountDisk:disk ?: self.diskRef];
-    if (mountURL) {
-      NSLog(@"%@ Remounted target %@ as %@", self, self.destination.path, mountURL.path);
-      [self applyImageInfo:mountURL];
-      [self blessMountURL:mountURL];
-      [self unmountDisk:disk ?: self.diskRef];
-    } else {
-      NSLog(@"%@ Unable to remount target, skipping imaginfo.plist application", self);
-    }
+  if (asrReturnCode != 0) {
+    NSError *err = [self constructErrorFromReturnCode:asrReturnCode];
+    [[self.client remoteObjectProxy] imageAppliedSuccess:NO error:err];
+    return;
+  }
 
+  // Get the synthesized "Macintosh HD" apfs disk, if any.
+  DADiskRef disk = NULL;
+  NSString *apfsOSDisk = [self apfsOSDisk];
+  if (apfsOSDisk) {
+    NSString *apfsBootDisk = [NSString stringWithFormat:@"/dev/%@", apfsOSDisk];
+    disk = DADiskCreateFromBSDName(NULL, self.diskArbSession, apfsBootDisk.UTF8String);
+  }
+  NSURL *mountURL = [self mountDisk:disk ?: self.diskRef];
+
+  // Fail if the disk would not mount and the post script is required.
+  // Succeed if the disk would not mount and the post script is not required.
+  // Otherwise continue on.
+  if (!mountURL && self.image.postScript && self.image.postScriptMustSucceed) {
+    NSString *s = @"Unable to remount target and the post script must succeed!";
+    NSDictionary *info = @{ NSLocalizedDescriptionKey: s };
+    NSError *err = [NSError errorWithDomain:@"com.google.corp.restord" code:777 userInfo:info];
+    [[self.client remoteObjectProxy] imageAppliedSuccess:NO error:err];
     if (disk) CFRelease(disk);
-  } else {
-    err = [self constructErrorFromReturnCode:asrReturnCode];
+    return;
+  } else if (!mountURL)  {
+    NSLog(@"%@ Unable to remount target, skipping imaginfo.plist and post script", self);
+    [[self.client remoteObjectProxy] imageAppliedSuccess:YES error:nil];
+    if (disk) CFRelease(disk);
+    return;
   }
 
-  // Send final reply with task status.
-  [[self.client remoteObjectProxy] imageAppliedSuccess:(asrReturnCode == 0) error:err];
+  NSLog(@"%@ Remounted target %@ as %@", self, self.destination.path, mountURL.path);
+
+  // Lay down the image info first so the post script can use the data.
+  [self applyImageInfo:mountURL];
+
+  // Run the post script.
+  // Fail if the post script failed and the post script is required.
+  if (self.image.postScript) {
+    [[self.client remoteObjectProxy] postScriptStarted];
+    NSError *err;
+    if ([self runPostScriptWithMountURL:mountURL error:&err] != 0 &&
```

Comment:
Is there any indication in Restor that the post script is running (or if it fails, that it failed, even if MustSucceed is false)?

Reply:
Yes, the UI shows when the script is running. It will also show an error if the script fails and the must succeed option is set. The error contains the stderr from the script.